### ISSUE TITLE
Make auto redeployment work for spring-boot health-check & crud quickstart

### ISSUE THREAD

**Issue by ibuziuk** [redhat-developer/rh-che#254]
Che version - 1.0.270
Demo - https://youtu.be/yY8sNJLqLko

Comment from @l0rd :
> @ibuziuk not sure what doesn't work here but maybe this comment may help: openshiftio/openshift.io#416 (comment). openshiftio/openshift.io#416 was closed but the problem was not fixed "entirely" and we still need to fix it. You may want to open a new issue about that in rh-che repo.

**Comment by ibuziuk** [comment]
@eivantsov could you please tell if hot deployment supposed to work for vert.x & spring-boot run / debug commands - https://youtu.be/yY8sNJLqLko

**Comment by l0rd** [comment]
@ibuziuk I pretty sure that hot deployment works fine for run command of vert.x stack

**Comment by eivantsov** [comment]
@ibuziuk @l0rd I think there's no Che fault here... do you expect the same behavior when running vetx locally?

**Comment by ibuziuk** [comment]
`@ibuziuk I pretty sure that hot deployment works fine for run command of vert.x stack`
@l0rd let me try again, but I was not able to make it work on osio
`@ibuziuk @l0rd I think there's no Che fault here... do you expect the same behavior when running vetx locally?`
@eivantsov I would expect at least debug commands to work with hot-swap (maybe I'm wrong)

**Comment by eivantsov** [comment]
@ibuziuk I'd try launching a vertx image with project sources mounted and debug ports exposed, and see if you can do it without Che...using any desktop IDE

**Comment by ibuziuk** [comment]
@l0rd @eivantsov I have tested redeployment of vertx app on osio and it seems to work, but for some reason the file that is edited is constantly updated in IDE - https://youtu.be/qr3fqaCix4E
Hot deployment works fine locally against the same project

**Comment by slemeur** [comment]
linking to https://github.com/openshiftio/openshift.io/issues/516

**Comment by slemeur** [comment]
@ibuziuk + @RomanNikitenko: Might be related to the upstream issue: https://github.com/eclipse/che/pull/5929

**Comment by eivantsov** [comment]
Yes, it's worth trying the latest Che. I cannot reproduce the issue with 1.0.270 build

**Comment by ibuziuk** [comment]
@eivantsov 1.0.270 does not contain the fix for eclipse/che#5929, and the problem is still reproducible for mee. We need to aline `openshift-connecot-wip` with master to pickup the fix

**Comment by ibuziuk** [comment]
Looks like hot-deployment for vet.x can't be used during debugging by design:
![image](https://user-images.githubusercontent.com/1461122/29632825-7ac0df8e-8844-11e7-8bca-730d677eb36d.png)
redeployment works just fine for vert.x run command though.
Also, it looks like spring boot run / debug commands are not supposed to support hot deployment. In order to enable hot swapping `spring-boot-devtools` should be configured in project (not configured in quick start) - https://docs.spring.io/spring-boot/docs/current/reference/html/using-boot-devtools.html

**Comment by eivantsov** [comment]
@ibuziuk yes, it's not Che fault here. Hot deployment when running vertx works ok for me.

**Comment by l0rd** [comment]
We should add spring-boot automatic redeployment when you select run command. To do this we should change global maven configuration.

**Comment by eivantsov** [comment]
But redeployment with run command works ok

**Comment by ibuziuk** [comment]
@eivantsov AFAIK it does not work for spring (works fine for vertx)

**Comment by eivantsov** [comment]
@ibuziuk should it work anyway for spring boot?

**Comment by l0rd** [comment]
More details about auto redeployment for spring-boot here: https://github.com/openshiftio/openshift.io/issues/416#issuecomment-318117751

**Comment by ibuziuk** [comment]
in order to make hot deployment work there should be a possibility of updating project's classpath [1] (In eclipse it is done automatically (`Project -> Build Automatically`) and in InteliJ one needs to either change a few build properties in IDE or manually run `Build -> Make Project`.
It looks like there is no equivalents for those operations which update classpath in Che. The best I was able to do is:

- add `spring-boot-devtools` dependency
- execute run command
- change the file
- execute build command which updates classpath
- SUCCESS: hot deployment works for run command 

![spring-boot-hot-deployment](https://user-images.githubusercontent.com/1461122/31444126-92e2983c-ae9b-11e7-9971-ac452797f57e.gif)

However, I do believe this is not the level of usability we are targeting.
@slemeur @eivantsov maybe I'm missing smth. and there is actually a possibility to enable classpath update on file change in Che? If no, WDYT about having an issue upstream for this functionality ? 
[1] https://docs.spring.io/spring-boot/docs/current/reference/html/using-boot-devtools.html#using-boot-devtools-restart

**Comment by l0rd** [comment]
@ibuziuk it seems to me that avoiding the stop/restart step is anyway a step forward (even if that is not ideal) and we should at least commit the changes to the stack to allow that.

**Comment by eivantsov** [comment]
@ibuziuk with auto save on, this can be a problem. We get catch file changes of course but do we need to update classpath for every maven project?

**Comment by ibuziuk** [comment]
> we should at least commit the changes to the stack to allow that.

IMO, this should not be done on the image level cause it is quite magical in a bad sense (installing dependency that is not presenting on project / having command which uses profile which is not specified in project), but rather done on quickstart project level. Also by doing this we will make stack commands image dependent 

> We get catch file changes of course but do we need to update classpath for every maven project?

@eivantsov what about InteliJ approach with `Build -> Make Project` ?

**Comment by eivantsov** [comment]
@ibuziuk does it differ much from rebuild? In terms of UX

**Comment by l0rd** [comment]
Ok let's update the spring boot quickstart then. And I would close this issue as soon as this is done, tracking the CLASSPATH update in another upstream issue.

**Comment by ibuziuk** [comment]
@eivantsov how to rebuild project in Che from UI ? :-)
or do you mean re-run build command ?

**Comment by eivantsov** [comment]
yes, that one. In terms of UX it looks the same to me. Project - Make or Commands - Build.

**Comment by ibuziuk** [comment]
well, Project -> Make Project is more straight forward IMO. But agree, those two approaches are quite similar.

**Comment by l0rd** [comment]
@ibuziuk can we close this one?

**Comment by ibuziuk** [comment]
@l0rd https://github.com/snowdrop/spring-boot-http-booster/pull/61 is merged so redeployment would work the following way for  `spring-boot-http-booster`. But I think we need to add the dependency to all other spring quickstarts 

[1] https://github.com/redhat-developer/rh-che/issues/254#issuecomment-335816550

**Comment by l0rd** [comment]
Alredy fixed for the HTTP booster. We need to update spring-boot-health-check and spring-boot-crud boosters too.

**Comment by ibuziuk** [comment]
PRs for `health-check` and `crud` quickstarts have been sent:
- https://github.com/snowdrop/spring-boot-health-check-booster/pull/17
- https://github.com/snowdrop/spring-boot-crud-booster/pull/19

**Comment by ibuziuk** [comment]
PRs above are meged - closing
